lib/galaxy/tool_util/deps/container_resolvers/explicit.py

```diff
@@ -108,7 +115,7 @@ def resolve(
                 return container_description
             image_id = cast(str, container_description.identifier)
             cache_path = os.path.normpath(os.path.join(self.cache_directory_path, image_id))
-            if install and not os.path.exists(cache_path):
+            if (self.auto_install or install) and not os.path.exists(cache_path):
```

Comment:
Why is install true in the first place for a resolver that can't write to the cached path ? Are you sure this isn't about the non-cached explicit resolver ?

If you don't want to work through these details I'm totally fine with skipping this branch if the cache path is not writable.

Reply:
(Which to mirror @bernt-matthias' sentiment, is really a bug fix that I have no problem with applying to 24.2)

Reply:
I'm happy to retarget to dev, Main will run it soon enough (or I can cherry pick) anyway.

That said, I am very unsure about the correctness of my fix here so am happy for the suggestions. How do we determine what is writable?

Reply:
Filesystems being filesystems and lying about things like write access, the only way to know for sure is to attempt to write.

Regarding your original question - part of the issue is I am potentially still confused on what the cached and non-cached resolvers are supposed to do.

Reply:
Both `explicit_singularity` and `explicit` as I am using them do not in my usage pull in the handler, they pull at runtime, and I don't see where in the code they would pull? 

`mulled` and `mulled_singularity` are different from their analogous `explicit*` resolvers in this regard, which is a significant source of the confusion.

Reply:
The cached ones do if you flip the install flag. The non-cached ones probably should ? I never worked this through since we didn't use explicit containers in the beginning (in fact I never enabled the non-cached resolvers at all).

Reply:
Apart from this discussion, there's also my opening statement: why is `install` true for you ? that just seems wrong ?

Reply:
Do you mean `auto_install` or `install`? `install` is true because the default in `tool_util.deps.containers:ContainerRegistry.resolve()` is (was) true, which is why I changed it in this PR.

I made `auto_install` true to mirror the default on the other resolvers that have the option. There's no indication just by the definition of a `cached_*` resolver whether or not it's writable. Obviously I want to set `auto_install` false in my config, that's the whole point of this PR.

Maybe worthwhile, we did also discuss the potential difference between `auto_install` and true read-only here: https://github.com/galaxyproject/galaxy/issues/7348